### DISCUSSION TITLE
GDevelop project's main JSON file `gdVersion` property purpose

### DISCUSSION THREAD

**Discussion by malec-palec** [4ian/GDevelop#6324] [Technical questions about the codebase]
Hi,

I wonder what the purpose of the `gdVersion` property in any GDevelop project's main JSON file is?
See example:
![game_version](https://github.com/4ian/GDevelop/assets/2181873/14ac8be2-8083-4168-bdef-97c0237330d2)

What is it used for?
And why does it never change?

**Comment by 4ian** [comment]
Mostly outdated/deprecated. Used in the past to allow to warn against opening a newer version of a GDevelop project using an old GDevelop version. With auto-updates (both for desktop/mobile) and the web version (which is "evergreen"), this is mostly useless.

Note that this number is still the version number of the GDevelop "Core" library (the C++ classes). But this is largely irrelevant now.

**Comment by malec-palec** [comment]
Thanks for the clarification.

Do you have any objections to restoring this functionality?
> to warn against opening a newer version of a GDevelop project using an old GDevelop version.

or at least just updating the actual version every time a project is built.

This will help desktop users with auto-update disabled. Or users who want to intentionally build a project using the exact version of the editor it was created in.
Also, it suits well for automation pipelines, e.g. building projects on CI/CD.

> But this is largely irrelevant now.

If so, it won't hurt either.

**Comment by f0nar** [comment]
@4ian Hi! I had an issue around GDevelop version updating too.
I had custom extension that used `RuntimeGame::getSceneData` method. Few versions ago you changed it to `RuntimeGame::getSceneAndExtensionsData` so my extension crushed in runtime and I realized it so late...

So the functionality as 
> to warn against opening a newer version of a GDevelop project using an old GDevelop version

sounds awesome.

OR 

Can we at least initialize `gdjs.projectData.gdVersion` with correct version of GDevelop on which project was build last time? In this case I can lock initial version in my extension and compare it with current one. It would be really nice. Let me know what you think

**Reply by 4ian** [reply]
Sorry about that, we indeed broke API compatibility, which we usually don't do but we've been thinking nobody was using this internally :p 
I don't have any particular argument against updating this, though I'm unsure how useful it will be: what do you think would be the advantage of checking the version? Your extension would probably hard crash/block the game by telling "Hey you're on an unsupported version of GDevelop, please update the extension or wait for it to be updated?"

Or do you have something else in mind? :)

**Reply by arthuro555** [reply]
Maybe generally, since it happened a few times in the past as well that extensions broke (thinking mostly of the recolorizer extension), it might be interesting to try and make GDevelop more smart about this:
- The call stack could be used to infer if and what extension the error comes from, and the name of the function and the extension causing the crash could be displayed in the red pop-up
- GDevelop crashes could be catched and sent to the IDE via the debugger connection, that way:
  - They could then be automatically reported to the GDevelop team to investigate via the editor's Posthog sdk (I believe it does error tracking?)
  - The IDE could suggest disabling the faulty extensions temporarily, if one is detected to crash the game a lot (which can be done simply by replacing the generated extension code with no-op functions)
  - The crash error messages would be inspect-able in the debugger console
- GDevelop could build a blacklist API queried when the IDE starts with the IDE version, and the API would return a list of extensions versions known to be broken for that version of GDevelop. It would then disable the extensions temporarily, and pop-up a dialogue urging the user to update GDevelop or the extension, depending on the breakage

**Reply by f0nar** [reply]
> Your extension would probably hard crash/block the game by telling "Hey you're on an unsupported version of GDevelop, please update the extension or wait for it to be updated?"

Hi!) That is close to the idea I have. I was thinking about an extension that remembers the version of project once it was created and adds it to "white list". In runtime it checks if current gd version is in the "white list". If no it throws. Then I can track version, verify changes and if everything is okay add it to "white list". I even can add circumstances according to version... Just thoughts...

> I'm unsure how useful it will be

In this case there would be at least smth for me to do. Maybe it is not for everybody but it is small thing that nice to have. In addition I think it will not break anything for you but give you ability to do with it more in the future if it is needed

**Reply by f0nar** [reply]
@arthuro555 what if after some code changes it creates logical error instead of runtime one? IDE will not track it but with different version trackers or lockers I can specify and control behavior of my extension with different gd versions, make backup to last "white listed" version or smth else. I can choose and define flow by my own.
I mean it would be nice to have smart IDE but I think it is pretty complicated and smart flow that is not mass adopted. As for me I would like to start from smth small.

**Reply by f0nar** [reply]
@4ian what do you think? Maybe I can start with PR?)

**Reply by 4ian** [reply]
Let's go with a PR :) 
Whatever the use case, it would be better to have the gdevelop version actually sync'ed with the editor version.

**Reply by f0nar** [reply]
Hi!
I see that you have not been supporting runtime versioning for a three years already. As I understand this process was stopped with GDevelop 5 and js usage. I see that `VersionWrapper` and surrounding stuff is still there so probably the main part is to sync `libGD` version with `electron-app/app/package.json` one. Problem is that `libGD` should be ready and built before js part to satisfy correct pipe. My question is if you are bumping gd version by hand as i see a lot of ["Bump version"](https://github.com/4ian/GDevelop/commit/eaa0f4077fb15daba089fd81020ddec4b470b28c) commits or you are doing it by some kind of optimization? If it first case so I understand that there should be one additional rule as "bump version in VersionPriv.h too", right? I am talking only about approach and logic. I understand that there should be additional testing and changes.

**Reply by 4ian** [reply]
@f0nar sorry for the delay! We're manually bumping GDevelop version indeed by updating `newIDE/electron-app/app/package.json`. This is also read by this script: https://github.com/4ian/GDevelop/blob/master/newIDE/app/scripts/make-version-metadata.js which generates a JS file that allows the editor to get the full version (as well as one with the git commit hash).

So we need to either:
- expose this version number to the C++ (by updating `Core/GDCore/Tools/VersionPriv.h` with a script?). For example, a Node.js script that would be run before GDevelop.js compilation. This would be a new task in https://github.com/4ian/GDevelop/blob/master/GDevelop.js/Gruntfile.js
- do the reverse: read or update `newIDE/electron-app/app/package.json` by reading VersionPriv.h (with a script that reads it and update the file).
- Something else? But if possible, let's avoid forcing 2 files to be updated. It would be cumbersome and error prone :)

At the end, we want something that is:
1) as simple as possible (so that's why I'm open to reverse the way it works)
2) that allows us to bump the version number (whenever we make a new official release) by changing one number and merging the changes (and then everything gets updated correctly).

Let me know what you think. I'm open to store the version in C++ and have the other files updated by scripts at compilation/npm start/something else.
